File: C version/link/detectCycle.c
#include <stdio.h>
#include <stdlib.h>
struct ListNode {
    int val;
    struct ListNode *next;
};
struct ListNode *detectCycle(struct ListNode *head) {
     struct ListNode* slow = head;
    struct ListNode* fast = head;
    bool  hasCircle = false;
    for(;;){   
        if(slow!=NULL){
            slow = slow->next;
        }
        if(fast!=NULL){
            fast = fast->next;
            if(fast!=NULL){
                fast = fast->next;
            }
        }
        if(slow==NULL && fast==NULL){
            break;
        }
        if(slow == fast || (fast!=NULL && fast->next == slow)){
            hasCircle = true;
            break;
        }
    }
    return NULL;
}
void main(){

}

File: LinkedList/detectCycle.js
//Definition for singly-linked list.
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}
const ListEntity = require("./makeListFromArray");
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var detectCycle = function(head) {
    let set = new Set;
    for(;;){
        if(head===null){break;}
        if(set.has(head)){
            return head
        }
        set.add(head);
        head = head.next;
    }
    console.log(set);
    return null;
};
let l1 = ListEntity.makeListFromArray([-1,-7,7,-4,19,6,-9,-5,-2,-5]);
console.log(detectCycle(l1));